### PR TITLE
Basic skeleton for FFI

### PR DESCRIPTION
**Pull request by c410-f3r**
Usually, it is not so hard to work with pointers across FFI boundaries. It is hard to use a single code-base for different language exports, properly manage errors and avoid memory corruption or undefined behaviour.

In face of such challenges, the `ffi-support` crate is a great help. Maintained and used by some of the Mozilla projects, provides utilities for safe interactions and error handling.

After some thoughts, I think the overall architecture will look like this:

* Use `ffi-support` for some internal details.
* At least for now, prefer JNA over JNI.  JNI expects a pre-defined format and even with the `jni` crate, it will be necessary to create yet another set of interfaces.
* `cgindgen` to create the C/C++ headers.

It is still not clear how well the integration of these parts will perform in face of exporting to JAVA/Kotlin/C/C++ or other languages. Maybe the answer will be provided with some more work.

#### This PR is composed of 3 commits:

* `Replace Mnemonic with &str`: `&str` is cheap, doesn't require an etra heap allocation and is common in function signatures due to its generalization. Besides, `Mnemonic` is a wrapper around `String` that is currently being used in places where ownership is not required at all. This change do not touch places where an owned `String` is returned by a function call. 

* `Add FFI skeleton`: Just a basic infrastructure that is probably going to be used for the rest of the public API.

* `CI script to test and check leaks`: Calls the built-in Rust tests of the `filecoin-signer-ffi` crate and also applies valgrind in the same binary to check possible memory leaks. Publishing of FFI artifacts is still WIP. 

#### More references about this subject:

* https://www.ralphminderhoud.com/blog/rust-ffi-wrong-way/
* https://github.com/mozilla/application-services/blob/master/docs/howtos/exposing-rust-components-to-kotlin.md
* https://github.com/filecoin-project/rust-fil-proofs/issues/889

### CODE REVIEW

**Comment by jleni** [comment]
Great work! 
Can you change the PR target branch to `dev` for now?

**Comment by c410-f3r** [comment]
@jleni Changed the target branch

After some evaluation, experimentation and headaches with JNA, it is more clear now that a DSL might be a better approach to generate code for FFI and JNI. Yes, JNI (which is faster than JNA) with the `jni` crate.

Such DSL was written using declarative macros for a faster MVP although they can be translated into procedure macros in the future. An optional feature `with-jni` was also added to easily switch between the different interfaces.

One important detail that is worth watching is the fact that each API change must be manually reflected into the `Filecoin.java` file.

File: examples/ffi/kotlin/README.md [hunk 1]
```diff
@@ -0,0 +1,3 @@
+# Kotlin bindings
+
+Please, use the Java bindings to work with Kotlin and Filecoin.
```
Comment (c410-f3r) [inline examples/ffi/kotlin/README.md:3]: IMHO, it is not necessary to create a Kotlin example

File: examples/ffi/c++/README.md [hunk 2]
```diff
@@ -0,0 +1,3 @@
+# C++ bindings
+
+Please, use the C bindings to work with C++ and Filecoin.
```
Comment (c410-f3r) [inline examples/ffi/c++/README.md:3]: Same as above

File: .circleci/config.yml [hunk 3]
```diff
@@ -54,6 +54,60 @@ jobs:
...
 
+  tests_ffi:
+    docker:
+      - image: zondax/rust-ci:latest
+    steps:
+      - checkout
+      - restore-sccache-cache
+      - run:
+          name: Build, test and run everything related to FFI
+          command: |
```
Comment (c410-f3r) [inline .circleci/config.yml:65]: IMO, each language should have its own job in a separated file, say `ffi.yml`.

Comment (jleni) [reply .circleci/config.yml:65]: AFAIK there is yet no support for multiple separate configuration files...

Comment (jleni) [reply .circleci/config.yml:65]: AFAIK, Circleci has not yet support for multiple files...

I would suggest the following:
- move `sudo apt to its own `run` called dependencies
- create a small shell script `test_LANG` for each LANG is c, cpp, java, etc.
- create separate runs for each: `run: ./test_cpp` .. etc.
That way CI dashboard is easier to understand, and eventually the scripts can be run locally too
- create a `run: ./test_LANG` so they are not 
- split this in `run` statements so it is possible to quickly see which language failed in CI
- move the

**Comment by jleni** [comment]
Awesome work!

**Comment by c410-f3r** [comment]
I wasn't able to run the vanilla docker from the Fedora repository and the cm version doesn't provide support for version 32, thus, my attempts to fix the CI remotely.
Turns out it didn't end so well so I will try to install locally again to properly test before submitting another commit.

File: .circleci/config.yml (hunk 3 above)
Comment (c410-f3r) [reply .circleci/config.yml:65]: Maybe one statement is missing? the `* move the ...`

File: .circleci/config.yml [hunk 4]
```diff
@@ -54,6 +54,27 @@ jobs:
...
+      - run:
+          name: Initial configuration
+          command: |
+            sudo apt-get install -y default-jdk g++ gcc valgrind snapd
+            wget -O sdk.install.sh "https://get.sdkman.io"
+            sudo bash sdk.install.sh
+            sudo source "/root/.sdkman/bin/sdkman-init.sh"
+            sudo sdk install kotlin
+            rustup install nightly
+            cargo +nightly install cbindgen
```
Comment (c410-f3r) [inline .circleci/config.yml:72]: These dependencies could be provided by a custom development image

Comment (jleni) [reply .circleci/config.yml:72]: yes, I will open an issue for this https://github.com/Zondax/docker-builder/issues/2
we can prioritize that later

**Comment by jleni** [comment]
I will open an issue for fix CI issues
https://github.com/Zondax/filecoin-rs/issues/110

**Pull request merged**